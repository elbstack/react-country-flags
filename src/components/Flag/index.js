import React, { PureComponent as Component, createElement } from 'react';
import PropTypes from 'prop-types';
import flags from '../../flags/react';

const fetchFlag = async country => {
  if (!flags[country]) return Promise.resolve(null);
  const { default: data } = await flags[country]();
  return data;
};

// thanks to https://github.com/facebook/react/issues/5465#issuecomment-157888325
const makeCancelable = promise => {
  let hasCanceled_ = false;

  const wrappedPromise = new Promise((resolve, reject) => {
    promise.then(
      val => (hasCanceled_ ? reject({ isCanceled: true }) : resolve(val)),
      error => (hasCanceled_ ? reject({ isCanceled: true }) : reject(error))
    );
  });

  return {
    promise: wrappedPromise,
    cancel() {
      hasCanceled_ = true;
    },
  };
};

export const countries = (() => Object.keys(flags))();

export default class Flags extends Component {
  cancelUpdate = () => {};

  state = {
    flag: null,
  };

  static defaultProps = {
    country: null,
    asSquare: false,
  };

  componentDidMount() {
    this.updateFlag();
  }

  componentDidUpdate(prevProps) {
    const { country } = this.props;

    if (country !== prevProps.country) {
      this.updateFlag();
    }
  }

  componentWillUnmount() {
    this.cancelUpdate();
  }

  updateFlag() {
    const { country } = this.props;
    const { promise, cancel } = makeCancelable(fetchFlag(country));
    promise.then(flag => {
      this.setState({ flag });
    });
    this.cancelUpdate = cancel;
  }

  render() {
    const { asSquare, country, ...props } = this.props;
    const { flag } = this.state;
    if (!flag && !asSquare) return null;
    if (!asSquare) return createElement(flag, props);
    return (
      <svg viewBox="0 0 100 100" preserveAspectRatio="xMidYMid" {...props}>
        {!flag ? null : createElement(flag)}
      </svg>
    );
  }
}

Flags.propTypes = {
  country: PropTypes.string,
  asSquare: PropTypes.bool,
};
