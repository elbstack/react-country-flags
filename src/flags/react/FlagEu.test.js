import React from 'react';
import renderer from 'react-test-renderer';
import Flag from './FlagEu';

it('FlagEu renders without crashing', () => {
  const component = renderer.create(<Flag />);

  let tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});
