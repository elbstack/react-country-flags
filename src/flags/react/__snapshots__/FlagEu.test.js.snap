// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FlagEu renders without crashing 1`] = `
<svg
  viewBox="0 0 810 540"
  xmlnsXlink="http://www.w3.org/1999/xlink"
>
  <defs>
    <g
      id="s91bf8332"
    >
      <g
        id="sf412d9e4"
      >
        <path
          d="M0 0v1h.5z"
          id="s0f53adc6"
          transform="rotate(18 3.157 -.5)"
        />
        <use
          transform="scale(-1 1)"
          xlinkHref="#s0f53adc6"
        />
      </g>
      <g
        id="sa8efbd1a"
      >
        <use
          transform="rotate(72)"
          xlinkHref="#sf412d9e4"
        />
        <use
          transform="rotate(144)"
          xlinkHref="#sf412d9e4"
        />
      </g>
      <use
        transform="scale(-1 1)"
        xlinkHref="#sa8efbd1a"
      />
    </g>
  </defs>
  <path
    d="M0 0h810v540H0z"
    fill="#039"
  />
  <g
    fill="#fc0"
    transform="matrix(30 0 0 30 405 270)"
  >
    <use
      xlinkHref="#s91bf8332"
      y={-6}
    />
    <use
      xlinkHref="#s91bf8332"
      y={6}
    />
    <g
      id="s5cea22cb"
    >
      <use
        x={-6}
        xlinkHref="#s91bf8332"
      />
      <use
        transform="rotate(-144 -2.344 -2.11)"
        xlinkHref="#s91bf8332"
      />
      <use
        transform="rotate(144 -2.11 -2.344)"
        xlinkHref="#s91bf8332"
      />
      <use
        transform="rotate(72 -4.663 -2.076)"
        xlinkHref="#s91bf8332"
      />
      <use
        transform="rotate(72 -5.076 .534)"
        xlinkHref="#s91bf8332"
      />
    </g>
    <use
      transform="scale(-1 1)"
      xlinkHref="#s5cea22cb"
    />
  </g>
</svg>
`;
